### PR TITLE
nimPackages cross-compilation fix

### PR DESCRIPTION
**Pull request by ghost**
###### Description of changes

The Nim compiler can build for any supported target platform so disregard `targetPlatform` and build for `hostPlatform` instead.

The `nim_builder` is irrelevant for use within `buildNimPackage` and should therefore be taken from buildPackages.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mindavi** [comment]
I _think_ this is ok? Did you try cross-compiling any nim package with this change (both inside nixpkgs and outside)?

**Comment by ehmry** [comment]
I am also not sure if this is how it works, but I've tested cross-compiling x86_64-linux -> aarch64-linux and x86_64-linux -> x86_64-genode.

It would be nice if there was an automatic cross-compile test I could add something to.

**Comment by Mindavi** [comment]
@ofborg build nitter pkgsCross.aarch64-multiplatform.nitter

**Comment by ehmry** [comment]
@NixOS/exotic-platform-maintainers hey cross folks, does this look right?

**Comment by sternenseemann** [comment]
The compiler should be built for `hostPlatform`, sure, but it should produce code for the `targetPlatform` *always*, this is nixpkgs assumption. The target prefix should be `targetPlatform` and if nim is multi target, we should (at least for the purposes of building derivations in nixpkgs) add a wrapped nim that passes the appropriate flags to fix the current `targetPlatform` (like we do for clang).

File: pkgs/top-level/nim-packages.nix [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 
 lib.makeScope newScope (self:
   let callPackage = self.callPackage;
   in {
     inherit nim;
     nim_builder = callPackage ../development/nim-packages/nim_builder { };
     buildNimPackage =
-      callPackage ../development/nim-packages/build-nim-package { };
+      callPackage ../development/nim-packages/build-nim-package {
+        inherit (buildPackages.buildPackages.nimPackages) nim_builder;
```
Comment (ehmry) [inline pkgs/top-level/nim-packages.nix:10]: Always building for target makes sense, but `buildInputs` seems to grab packages from `buildPackages` but for `depsBuildBuild` I manually need to take packages from `buildPackages.buildPackages`?

Comment (symphorien) [reply pkgs/top-level/nim-packages.nix:10]: maybe you should use makeScopeWithSplicing

Comment (ehmry) [reply pkgs/top-level/nim-packages.nix:10]: What would be the difference?

Comment (symphorien) [reply pkgs/top-level/nim-packages.nix:10]: makeScope breaks splicing, ie the mechanism that makes `nativeBuildInputs = [ foo ]` and `buildInputs = [ foo ]` pick a different derivation.

Comment (ehmry) [reply pkgs/top-level/nim-packages.nix:10]: That sounds good, but `overrideScope'` then is missing?

Comment (symphorien) [reply pkgs/top-level/nim-packages.nix:10]: I don't think makeScopeWithSplicing is a drop-in replacement, but of course it's not documented...

**Comment by Ericson2314** [comment]
Yeah what @sternenseemann said. This needs to be reworked.

**Comment by ehmry** [comment]
Ok, I'll keep an eye on it.

**Comment by Ericson2314** [comment]
What huh? Unless I misread @sternenseemann was saying this is not done correctly yet and needs revision. That is what I meant to say at least.

**Pull request merged**